text/0028-generalized-entity-data-source.md

```diff
@@ -0,0 +1,569 @@
+        // which may be inefficient for some implementations.
+        Ok(self.entity_attr(uid, attr)?.is_some())
+    }
+
+    /// Is `child in parent`?
+    ///
+    /// Implementations of this method should not assume that either entity exists.
+    /// If either entity does not exist, the implementation should return
+    /// `Err(UnknownEntity)` and not `Ok(false)`.
+    fn entity_in(&self, child: &EntityUid, parent: &EntityUid) -> Result<bool>;
+}
+```
+
+For convenience, to support the common case where users want to load entity data
+in units of entire entities, we also provide a trait `WholeEntityDataSource`
+(open to other name suggestions):
+
+```rust
+pub trait WholeEntityDataSource {
+    /// Given a uid, get the corresponding entity if it exists.
+    /// The entity should already have `ancestors` data which represents all
+    /// transitive closure ancestors, not just the direct parents.
+    fn get(&self, uid: &EntityUid) -> Result<Option<Entity>>;
+}
+```
+
+_Note: we could have `get()` return a `Result<Option<Arc<Entity>>>` or `Result<Option<Cow<Entity>>>` in order to avoid unnecessary copying of `Entity` in some implementations. We consider details like this to be implementation details that can be handled at the implementation stage and don't need to be discussed or decided in this RFC._
+
+Then, we provide a blanket implementation of `EntityDataSource` for every type
+`T` that implements `WholeEntityDataSource`: (pseudocode shown)
+```rust
+impl<T: WholeEntityDataSource> EntityDataSource for T {
+    fn entity_exists(&self, uid: &EntityUid) -> Result<bool> {
+        // computes the entire entity with `.get()`, only to throw it away,
+        // which is inefficient in implementations where `.get()` is expensive.
+        Ok(self.get(uid)?.is_some())
+    }
+    fn entity_attr(&self, uid: &EntityUid, attr: &str) -> Result<Option<Value>> {
+        // computes the entire entity with `.get()`, only to throw it away,
+        // which is inefficient in implementations where `.get()` is expensive.
+        Ok(self.get(uid)?.attr(attr))
+    }
+    fn entity_in(&self, child: &EntityUid, parent: &EntityUid) -> Result<bool> {
+        // computes the entire entity with `.get()`, only to throw it away,
+        // which is inefficient in implementations where `.get()` is expensive.
+        Ok(self.get(child)?.in(parent))
+    }
+}
+```
+
+We also provide the trivial implementation of `WholeEntityDataSource` for
+`Entities`.
+This allows users to continue passing an `Entities` to provide
+entity data to `is_authorized()`, and makes the change to `is_authorized()`
+backwards-compatible: it still accepts an `Entities`, but now also accepts
+additional types, namely any other `EntityDataSource`.
+Note that the implementation for `Entities` doesn't suffer from the potential
+inefficiencies noted in the code comments above: `.get()` is `O(1)` and just
+returns a reference to an existing `Entity` already stored in-memory.
```

Comment:
I'm now wondering whether we actually want to expose `WholeEntityDataSource` to our users, or only use it internally to provide the impl of `EntityDataSource` for `Entities`.  In what situation would it be the correct choice for users to provide a custom implementation of `WholeEntityDataSource`?  If their implementation of `.get()` is expensive, it will be very inefficient as noted here, because Cedar will call `.get()` potentially many times for the same entity and throw away the result each time.  But if their implementation of `.get()` is cheap, constructing an `Entities` would be cheap, and perhaps we should encourage them to do that instead?  (and just use the `Entities` implementation of `WholeEntityDataSource`.)  

Or perhaps we should redesign our API so that the Cedar evaluator can request a whole-entity but actually remember the entity data instead of throwing it away, and use that instead of calling `.get()` again in the future.  Basically the Cedar evaluator caching the results of calling `.get()`.  Even if the cache only lasts the length of a single `is_authorized()` call and is thrown away when the `Response` is returned.

Reply:
This reduces the motivation for considering all four components of this RFC at once, because one reason for that was to help users (efficiently) implement `WholeEntityDataSource`; see [this thread](https://github.com/cedar-policy/rfcs/pull/28#discussion_r1349244359).  Perhaps we'll need to sever some components of this RFC into a new RFC.  I will leave it as one RFC for now though.

Reply:
`WholeEntityDataSource` is now removed from the discussion in the RFC.